Adding vm-templates design (#173)
openshift/openshift-origin-design: web-console/knikubevirt/vm-templates/vm-templates.md

# Virtual machine templates

### Summary

> - VM-templates is located at the main side menu, in 'Workloads' under 'virtual machines'.

Reviewer: for consistency can we have this say "Virtual Machine Templates" are located under "Virtual Machines" in the primary navigation.

> - VM-templates displays a list view similar to VM list with minor changes in columns and actions.

Reviewer: ```suggestion
- VM-templates displays a list view similar to the VM list view with relevant changes in columns and actions.
```

> - From this List, the user can create a new template, create a virtual machine from a selected template, edit a selected template, or delete a selected template.

Reviewer: ```suggestion
- From this list the user can create a new template, create a virtual machine from a template, edit a template, or delete a template.
```

Reviewer: nit pic but this reads a bit better.

- Clicking on a template will bring the user to the template’s Overview tab (TBD).
Creating or editing a template will display a wizard similar to the “No Template state” wizard from Create VM but with slight adjustments.
- Users can also create templates based on existing VMs from the actions menu of any VM within the VM List View.

## Template list view

![Offline Template List View](img/1-0-vm-template-list.jpg)

> The user will be able to view a list of all templates that exist. From this List, the user can create a new template, create a virtual machine from a selected template, edit a selected template, or delete a selected template.

Reviewer: same changes as above?

## Create template from primary button

![Offline Template List View](img/1-1-0-create-template.jpg)

Clicking “Create Template” will display a wizard similar to the “No Template state” wizard from Create VM but with slight adjustments.

![Offline Template List View](img/1-1-1-create-template-modal.jpg)

1. Modal title changes to “Create Template”

> 2. “Template” is no longer a Provision Source option

Reviewer: This design has been updated to show Templates as a separate dropdown in the create a VM wizard. This could read: "Choosing a template from a dropdown is not available" or something like that.

4. The quick action button at the bottom is now “Create Template”
5. In Step 5 “Review”, references to “Virtual Machine” are replaced with “Template”

    a. Create/Creating Virtual Machine -> Create/Creating Template

    b. Virtual Machine creation failed -> Template creation failed

    c. Virtual machine created successfully -> Template created successfully

Everything else remains the same.

## Create template from VM List

![Offline Template List View](img/1-2-0-create-template-from-vm-list.jpg)

Users can also create templates based on existing VMs from the actions menu of any VM within the VM List View. 

![Offline Template List View](img/1-2-1-create-template-from-vm-list-modal.jpg)

The VM they choose pre-fills the Provision Source, OS, Flavor, Workload Profile, and all other Networking, Storage, and Advanced settings of the Create Template wizard.

## Create VM from another template

![Offline Template List View](img/1-3-0-create-template-from-template-list.jpg)

Users can initiate the Create VM wizard from the actions menu of a VM template.

![Offline Template List View](img/1-3-1-create-template-from-template-list-modal.jpg)

> The template they choose is preselected and the OS, Flavor, and Workload Profiles are filled in. Refer to the Create VM wizard for all other steps.

Reviewer: we might want to drop the line "Refer to the Create VM wizard for all other steps" line since it's not available yet.

> The template they choose is preselected and the OS, Flavor, and Workload Profiles are filled in. Refer to the Create VM wizard for all other steps.

Reviewer: ```suggestion
The template they choose is preselected and the Provision Source, OS, Flavor, and Workload Profiles are filled in. Refer to the Create VM Wizard for all other steps.
```

![Offline Template List View](img/1-2-2-create-template-from-vm-list-starting.jpg)

After successfully working through the Create VM Wizard, the user would be taken to the Virtual Machine List View to show the new VM creation progress.

## Template details (coming soon)

Clicking on a template will bring the user to the template’s Overview tab.

> The YAML tab allows the user to edit the template’s configuration.

Reviewer: Is this still the case @lizsurette, will it be the same as how the edit state works for the VM details inline?

Reviewer 2: Yes this is implemented today, the user can edit the YAML for both VMs as well as VM Templates.

Reviewer: ya, we also have the additional editing via the ui, would it be the same for here?

Reviewer 2: Yeah that makes total sense to me to allow the user to edit the template via a nice form as well.

Author: Adding this to the .md

> The Virtual Machines tab displays a VM List View with only the VMs that were created using the template. This list behaves mostly the same, but the actions kebab will likely need to be different (no Create Template button, for example)

Reviewer: should we say "behaves the same with the relevant actions available in the kebab". It is a bit less open ended.

Author: Changed that according to your suggestion.
This list relates to a discussion I created in the doc, which I find a bit off:

- I suspect that this list suggests that there is a relation between a template and a child. Because of this suggestion, I misunderstood the doc and thought that it suggests creating a link between the template and the VM created from it (making it so that any changes applied on the template will change the VM). 
- I'm not sure what kind of action will come out of viewing this list.
- Since a template is a mere starting point for creating a VM, some or all VMs created from it will be modified. So, will they be visible here? what're the modifications, threshold?

Reviewer: Ya that's a good point. If a VM is created from a template but then is modified would we still want to list that vm as a result of that template? @lizsurette how much do you recall about this design, do you know why we are listing the VMs that were created from the template? It would be good to know the use case.

Reviewer 2: It's probably worth a follow-up with Moran, but we definitely went back and forth between there being a strong relationship between VMs and Templates vs. not doing that. Eventually we decided to start without that strong relationship. Maybe this could be removed? Or maybe made more clear that these VMs were created based on this template, but may have changed a bit since initial creation.

## Editing a template (coming soon)

> Likely brings up the same creation wizard, but with any mention of “Create” changed to “Save” instead.

Reviewer: Can this be more declarative? The word "likely" sounds too open. If we aren't sure yet maybe we should remove this until we know.

Author: It's related to the discussion between you and Liz, so I'm just removing "Likely".

# Virtual machine templates

### Summary

- "Virtual Machine Templates" are located under "Virtual Machines" in the primary navigation.
- VM-templates displays a list view similar to the VM list view with relevant changes in columns and actions.
- From this list the user can create a new template, create a virtual machine from a template, edit a template, or delete a template.
- Clicking on a template will bring the user to the template’s Overview tab (TBD).
Creating or editing a template will display a wizard similar to the “No Template state” wizard from Create VM but with slight adjustments.
- Users can also create templates based on existing VMs from the actions menu of any VM within the VM List View.

## Template list view

![Offline Template List View](img/1-0-vm-template-list.jpg)

The user will be able to view a list of all templates that exist. From this list the user can create a new template, create a virtual machine from a template, edit a template, or delete a template.

## Create template from primary button

> ![Offline Template List View](img/1-1-0-create-template.jpg)

Author: Agree, will remove the description for now

Clicking “Create Template” will display a wizard similar to the “No Template state” wizard from Create VM but with slight adjustments.

![Offline Template List View](img/1-1-1-create-template-modal.jpg)

1. Modal title changes to “Create Template”
2. Choosing a template from a dropdown is not available
3. The two checkboxes for “Start virtual machine on creation” and “Create new template from configuration” are removed
4. The quick action button at the bottom is now “Create Template”
5. In Step 5 “Review”, references to “Virtual Machine” are replaced with “Template”

    a. Create/Creating Virtual Machine -> Create/Creating Template

    b. Virtual Machine creation failed -> Template creation failed

    c. Virtual machine created successfully -> Template created successfully

Everything else remains the same.

## Create template from VM List

![Offline Template List View](img/1-2-0-create-template-from-vm-list.jpg)

Users can also create templates based on existing VMs from the actions menu of any VM within the VM List View. 

> ![Offline Template List View](img/1-2-1-create-template-from-vm-list-modal.jpg)

Reviewer: Hey @beanh66 I’ve got an updated wizard on the way. Once we get that in we can move everything over. I have it as a modal wizard but don’t feel strongly about it. We could move it to in page if that’s going to be the pattern everywhere else in OpenShift.

The VM they choose pre-fills the Provision Source, OS, Flavor, Workload Profile, and all other Networking, Storage, and Advanced settings of the Create Template wizard.

## Create VM from another template

![Offline Template List View](img/1-3-0-create-template-from-template-list.jpg)

Users can initiate the Create VM wizard from the actions menu of a VM template.

![Offline Template List View](img/1-3-1-create-template-from-template-list-modal.jpg)

The template they choose is preselected and the Provision Source, OS, Flavor, and Workload Profiles are filled in. Refer to the Create VM Wizard for all other steps.

> ![Offline Template List View](img/1-2-2-create-template-from-vm-list-starting.jpg)

Reviewer 3: Same comment as pervious PRs. Not sure we need to tell users this has started if the action was just taken by the user and the status changes. Seems like enough feedback.

Author: @beanh66 
I agree that this behavior is not ideal, especially when we land the user on a VM list:

- Using a modal, we don't want to move the user to another page on 'Applying' that modal, but to return to the page in the background. Plus, there is no benefit of landing on the VM list. If it was the new VM, it would be a different case.
- When 'Applying' the modal, the user does need feedback to know that something happened.  But is enough to display the toast notification as a 'success' feedback, and add a link to 'View new VM' page. There is no need to land her out of context.

WDYT?

@matthewcarleton this 'landing on a different page from applying modal' also appears in other places, like VM snapshots. We need to align whatever we decide here to those other places as a rule of thumb

Author: I agree.
I think that sending the user to another page after 'applying' a modal is an issue altogether. The user should return to the background page.
On returning to the list view page, the user can view the new item on the list.
It is not solid feedback since we don't highlight that new item, but it is enough to remove the toast.

Author: I agree on removing the toast.

Sending the user to another page after 'applying' a modal is an issue. The user should return to the background page. On returning to the list view page, the user can view the new item on the list.
It is not solid feedback since we don't highlight that new item, but it is enough to remove the toast.

@matthewcarleton this 'landing on a different page from applying modal' also appears in other places, like VM snapshots. We need to align whatever we decide here to those other places as a rule of thumb
